doc/cpp-style-guide.md

```diff
@@ -0,0 +1,933 @@
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Preincrement_and_Predecrement)
+
+#### Use of const
+
+Use `const` whenever it makes sense. With C++11, `constexpr` is a better choice for some uses of `const`.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Use_of_const)
+
+**Where to put the const**:  we have been putting the const to the right as in `std::string const&`.  It is easy to read
+C++ types like this from right to left ("a reference to a constant string").
+
+#### Use of constexpr
+
+Use `constexpr` to define true constants or to ensure constant initialization.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Use_of_constexpr)
+
+#### Integer Types
+
+Use `<cstdint>` over `<stdint.h>`, qualify the integer types as `std::int16_t`.   The libraries in `google-cloud-cpp`
+will be used by external users, who may have more restrictive rules about polluting the global namespace
+
+This is substantially different from the corresponding
+[GSG section](https://google.github.io/styleguide/cppguide.html#Integer_Types)
+
+#### 64-bit Portability
+
+Code should be 64-bit and 32-bit friendly. Bear in mind problems of printing, comparisons, and structure alignment.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#64-bit_Portability)
+
+#### Preprocessor Macros
+
+Avoid defining macros, especially in headers; prefer inline functions, enums, and const variables. Name macros with a
+project-specific prefix. Do not use macros to define pieces of a C++ API.
+
+Macros mean that the code you see is not the same as the code the compiler sees. This can introduce unexpected
+behavior, especially since macros have global scope.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Preprocessor_Macros)
+
+#### 0 and nullptr/NULL
+
+Use `0` for integers, `0.0` for reals, `nullptr` for pointers, and `'\0'` for chars.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#0_and_nullptr/NULL)
+
+#### sizeof
+
+Prefer sizeof(varname) to sizeof(type).
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#sizeof)
+
+#### auto
+
+Use auto to avoid type names that are noisy, obvious, or unimportant - cases where the type doesn't aid in clarity
+for the reader. Continue to use manifest type declarations when it helps readability.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#auto)
+
+#### Braced Initializer List
+
+You may use braced initializer lists.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Braced_Initializer_List)
+
+#### Lambda Expressions
+
+Use lambda expressions where appropriate. Prefer explicit captures when the lambda will escape the current scope.
+
+[link to GSG](https://google.github.io/styleguide/cppguide.html#Lambda_expressions)
+
+#### Template Metaprogramming
+
+Avoid complicated template programming.
```

Comment:
The template metaprogramming section can be skipped and tackled on a commit by commit basis because "complicated" is subjective and will have to be dealt on a case by case basis.

Reply:
Agree, though I think I would prefer to keep this.  I do not want to deviate from the Google Style Guide where possible, and this rule basically has the same effect as "If we cannot understand your template meta-programming you need to change the code".

Reply:
Will defer to you on this.